### PR TITLE
Documentation: add document on the dex API

### PR DESCRIPTION
**Pull request by ericchiang**

### CODE REVIEW

**Comment by ericchiang** [comment]
@ElijahCaine ptal

File: Documentation/api.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  # Cannot be the same address as an HTTP(S) service.
+  addr: 127.0.0.1:5557
+  # Server certs. If TLS credentials aren't provided dex will generate self-signed ones.
+  tlsCert: /etc/dex/grpc.crt
+  tlsKey: /etc/dex/grpc.key
+```
+
+## Generating clients
+
+gRPC is a suite of tools for generating client and server bindings from a common declarative language. The canonical schema for dex's API can be found in the source tree at `api/api.proto`. Go bindings are generated and maintained in the same directory for internal use.
```
Comment (ElijahCaine) [inline Documentation/api.md]: I would make `api/api.proto` a link to the file on GitHub for convenience.
Most readers will expect that reference to be linked.

File: Documentation/api.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+To generate a client for your own project install [`protoc`][protoc], install a protobuf generator for your project's language, and download the `api.proto` file. An example for a Go project:
+
+```
+# Install protoc-gen-go.
```
Comment (ElijahCaine) [inline Documentation/api.md]: If these are shell commands they should begin with `$ `. If it's a script make that clear. I like annotating it like this:

`script-name.sh`:
```
# script contents
...
```

But anything that makes it clear works for me.

File: Documentation/api.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+# Download api.proto for a given version.
+DEX_VERSION=v2.0.0-alpha.5
+wget https://raw.githubusercontent.com/coreos/dex/${DEX_VERSION}/api/api.proto 
+
+# Generate the Go client bindings.
+protoc --go_out=import_path=dexapi:. api.proto
+```
+
+Client programs can them be written using the generated code. A Go client which uses dex's internally generated code might look like the following:
```
Comment (ElijahCaine) [inline Documentation/api.md]: Client programs can **then** be written [...]

File: Documentation/api.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+While the dex team would be open to re-implementing `dexctl` for v2 a majority of the work is writing a design document, not the actual programming effort.
+
+## Why not REST or gRPC Gateway?
+
+Between v1 and v2, dex switched from REST to gRPC. This largely stemmed from problems generating documentation, client bindings, and server frameworks that adequately expressed REST semantics. While [Google APIs][google-apis], [Open API/Swagger][open-api], and [gRPC Gateway][grpc-gateway] were evaluated, they often became clunky when trying to use specific HTTP error codes or complex request bodies. As a result, v2's API is entirely gRPC.
+
+Many of arguments _against_ gRPC cite short term convenience rather than production use cases. Though this is a recognized shortcoming, dex already implements many features for developer convenience. For instance, users who wish to manually edit clients during testing can use the `staticClients` config field instead of the API.
```
Comment (ElijahCaine) [inline Documentation/api.md]: Many arguments _against_ (get rid of `of`).

**Comment by ericchiang** [comment]
changes addressed. thanks

**Pull request merged**